reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServer.java

```diff
@@ -955,7 +956,9 @@ public void onStateChange(Connection connection, State newState) {
 						log.debug(format(connection.channel(), "Handler is being applied: {}"), handler);
 					}
 					HttpServerOperations ops = (HttpServerOperations) connection;
-					Mono<Void> mono = Mono.fromDirect(handler.apply(ops, ops));
+					Mono<Void> mono = Mono.deferContextual(Mono::just)
+					                      .doOnNext(ctx -> ops.currentContext = Context.of(ctx))
+					                      .then(Mono.fromDirect(handler.apply(ops, ops)));
```

Comment:
I would avoid extra couple operators (doOnNext, then) by setting the `ops.currentContext` and returning the `fromDirect` inside the `deferContextual` lambda.

Reply:
as discussed, this suggestion doesn't really work because the `catch` clause below depends on `handler#apply` throwing at assembly time. applying the function outside of `deferContextual` should work though.

Reply:
fixed